File: 1218-lowest-common-ancestor-of-deepest-leaves/1218-lowest-common-ancestor-of-deepest-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
   public:
    bool isLeaf(TreeNode *root) { return !root->left && !root->right; }

    pair<TreeNode *, int> solve(TreeNode *root, int depth) {
        if (!root) return {nullptr, 0};
        if (isLeaf(root)) return {root, depth};

        auto l = solve(root->left, depth + 1);
        auto r = solve(root->right, depth + 1);

        if (l.second > r.second) {
            return l;
        } else if (l.second < r.second) {
            return r;
        } else {
            return {root, l.second};
        }
    }
    TreeNode *lcaDeepestLeaves(TreeNode *root) { return solve(root, 0).first; }
};

File: 1218-lowest-common-ancestor-of-deepest-leaves/1218-lowest-common-ancestor-of-deepest-leaves.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isLeaf(root *TreeNode) bool {
	return root.Left == nil && root.Right == nil
}

func solve(root *TreeNode, depth int) (*TreeNode, int) {
	if root == nil {
		return nil, -1
	}
	if isLeaf(root) {
		return root, depth
	}

	l, ld := solve(root.Left, depth+1)
	r, rd := solve(root.Right, depth+1)

	if ld > rd {
		return l, ld
	} else if ld < rd {
		return r, rd
	} else {
		return root, ld
	}
}

func lcaDeepestLeaves(root *TreeNode) *TreeNode {
	ans, _ := solve(root, 0)
	return ans
}
